### PR TITLE
Fix: #3186

### PR DESCRIPTION
**Pull request by robert00s**
Close: #3186 (Statistics settings get reset to default when you just save them)

### CODE REVIEW

**Comment by Frenzie** [comment]
It's not immediately obvious to me what's happening. Why is it possible for `fields` to be true but fields[1] and/or fields[2] to be an empty string?

**Comment by robert00s** [comment]
When we save settings with empty values fields look that:

```
fields = {
[1] = "",
[2] = ""
}
```
so fields are `true` 
I this case fields are always true so we need to check `fields[1]` and `fields[2]` :)

**Comment by Frenzie** [comment]
But why are they empty as opposed to `self.page_min_read_sec == tonumber(fields[1])`?

**Comment by robert00s** [comment]
`tonumber(fields[1])` return nil when fields[1] is non number value.
I changed the code just a moment ago.

**Comment by Frenzie** [comment]
Oh, I see. The value is nil because there's a placeholder explanation in there. It didn't quite register before, but it brings to mind this warning from the HTML5 spec:

> Note: Use of the placeholder attribute as a replacement for a label can reduce the accessibility and usability of the control for a range of users including older users and users with cognitive, mobility, fine motor skill or vision impairments. While the hint given by the control's label is shown at all times, the short hint given in the placeholder attribute is only shown before the user enters a value. Furthermore, placeholder text may be mistaken for a pre-filled value, and as commonly implemented the default color of the placeholder text provides insufficient contrast and the lack of a separate visible label reduces the size of the hit region available for setting focus on the control.

Pinging @baskerville for an opinion.

screenshot_2017-09-08_22-22-11

**Comment by baskerville** [comment]
I'd go with *Min seconds* and *Max seconds* as labels and fill each field with its value (I think the user only wants to know what the value is and doesn't care about implementation details). *Apply* should be disabled if the values haven't changed.

**Comment by Frenzie** [comment]
If you don't mention the default value there should probably be a reset to
default button.

**Comment by baskerville** [comment]
> If you don't mention the default value there should probably be a reset to default button.

Probably.

**Comment by robert00s** [comment]
Something like that? :)
screenshot 2017-09-12 19-26-39
But I don't know how to disable Apply button and enable it after entering the values.

**Comment by Frenzie** [comment]
Yes, except I don't think it makes all that much sense to say "current value" as a hint rather than just sticking it in there. Plus that would remove the need for this fix.

I don't think the Apply button thing is currently easily possible. Something else to add to the mental todo list. :-)

File: plugins/statistics.koplugin/main.lua [hunk 1]
```diff
@@ -559,15 +559,15 @@ function ReaderStatistics:updateSettings()
...
             {
                 text = "",
+                description = T(_("Min seconds, default is %1"), DEFAULT_MIN_READ_SEC),
                 input_type = "number",
-                hint = T(_("Min seconds, default is %1. Current value: %2"),
-                    DEFAULT_MIN_READ_SEC, self.page_min_read_sec),
+                hint = self.page_min_read_sec,
             },
             {
                 text = "",
```
Comment (Frenzie) [inline plugins/statistics.koplugin/main.lua]: I just meant here without any hint. Unless you have any objections?

File: plugins/statistics.koplugin/main.lua [hunk 2]
```diff
@@ -1238,8 +1238,14 @@ end
 
 function ReaderStatistics:saveSettings(fields)
     if fields then
-        self.page_min_read_sec = tonumber(fields[1])
-        self.page_max_read_sec = tonumber(fields[2])
+        local fields1 = tonumber(fields[1])
+        local fields2 = tonumber(fields[2])
+        if fields1 then
```
Comment (Frenzie) [inline plugins/statistics.koplugin/main.lua]: Because then this isn't necessary at all.

File: plugins/statistics.koplugin/main.lua [hunk 3]
```diff
@@ -559,15 +559,15 @@ function ReaderStatistics:updateSettings()
         fields = {
             {
                 text = "",
```
Comment (poire-z) [inline plugins/statistics.koplugin/main.lua]: I think you could use `text = self.page_min_read_sec,`(what you put into `hint`), and it should be there as pre-filled value and editable, and you'll get it back if the user applies changes.

**Comment by Frenzie** [comment]
Awesome, thanks! Feel free to squash merge it yourself. :-)

**Comment by robert00s** [comment]
@Frenzie 
I think, I don't have permissions and I don't know how to do it :-)

**Comment by Frenzie** [comment]
Just above where you write a reply you should have a merge button with an arrow on the side of it in any PR.

**Pull request merged**